### PR TITLE
histogram

### PR DESCRIPTION
**Pull request by a311916000**
histogram graphs

### CODE REVIEW

File: interface/components/graph/index.coffee [hunk 1]
```diff
@@ -9,28 +9,209 @@ ko.components.register "tf-graph",
...
+          data_predicted = []
+          for i in [0...data.length]
+            data_predicted[i] = data[i][0]
+          data_predicted.sort (a, b) -> a - b
+          
+          min = data_predicted[0]
+          max = data_predicted[data_predicted.length - 1]+1
+          bucket_size = 10
+          buckets_width = (max-min)/bucket_size
+          console.log("data_predicted")
```
Comment (tsangjustin) [inline interface/components/graph/index.coffee]: Remove console logs

File: interface/components/graph/index.coffee [hunk 2]
```diff
@@ -9,28 +9,209 @@ ko.components.register "tf-graph",
...
+            hist_labels[i] = n
+          hist_labels.push(max.toFixed(2))
+
+          buckets = []
+          for i in [0...bucket_size]
+            buckets[i] = 0
+          for i in [0...data.length]
+            counter = Math.floor((data_predicted[i] - min) / buckets_width)
+            buckets[counter] = buckets[counter]+1
+          console.log("bucket within function")
```
Comment (tsangjustin) [inline interface/components/graph/index.coffee]: Remove this also

File: interface/components/graph/index.coffee [hunk 3]
```diff
@@ -9,28 +9,209 @@ ko.components.register "tf-graph",
...
+      get_hist_labels = (data) =>
+          result = []
+          data_predicted = []
+          for i in [0...data.length]
+            data_predicted[i] = data[i][0]
+          data_predicted.sort (a, b) -> a - b
+          
+          min = data_predicted[0]
+          max = data_predicted[data_predicted.length - 1]+1
+          console.log("min, max within hist labels:")
```
Comment (tsangjustin) [inline interface/components/graph/index.coffee]: Remove console log here

File: interface/components/graph/index.coffee [hunk 4]
```diff
@@ -9,28 +9,209 @@ ko.components.register "tf-graph",
...
+          console.log(min)
+          console.log(max)
+          bucket_size = 10
+          buckets_width = (max-min)/bucket_size
+          
+          for i in [0...bucket_size]
+            n = (min + (i*buckets_width)).toFixed(2)
+            console.log("i: " + i)
```
Comment (tsangjustin) [inline interface/components/graph/index.coffee]: here too

File: interface/components/graph/index.coffee [hunk 5]
```diff
@@ -9,28 +9,209 @@ ko.components.register "tf-graph",
...
+              return [data]
+          else if data_length == 4
+              listone = []
+              listtwo = []
+              for i in [0...data.length]
+                if(data[i][0] != null && data[i][1] != null)
+                    listone[i] = [data[i][0], data[i][1]]
+                if(data[i][2] != null && data[i][3] != null)
+                    listtwo[i] = [data[i][2], data[i][3]]
+              console.log("result")
```
Comment (tsangjustin) [inline interface/components/graph/index.coffee]: here too

File: interface/components/graph/index.coffee [hunk 6]
```diff
@@ -9,28 +9,209 @@ ko.components.register "tf-graph",
...
+              listthree = []
+              for i in [0...data.length]
+                if(data[i][0] != null && data[i][1] != null)
+                    listone[i] = [data[i][0], data[i][1]]
+                if(data[i][1] != null && data[i][2] != null)
+                    listtwo[i] = [data[i][2], data[i][3]]
+                if(data[i][4] != null && data[i][5] != null)
+                    listthree[i] = [data[i][4], data[i][5]]
+                
+              console.log("result")
```
Comment (tsangjustin) [inline interface/components/graph/index.coffee]: here

File: interface/components/graph/index.coffee [hunk 7]
```diff
@@ -9,28 +9,209 @@ ko.components.register "tf-graph",
...
+      buckets = []
+      for i in [0...bucket_size]
+        buckets[i] = 0
+      for i in [0...data().length]
+        counter = Math.floor((data_predicted[i] - min) / buckets_width)
+        buckets[counter] = buckets[counter]+1
+
+      
+      
+      console.log("buckets_width: ")
```
Comment (tsangjustin) [inline interface/components/graph/index.coffee]: here

File: interface/components/graph/index.coffee [hunk 8]
```diff
@@ -9,28 +9,209 @@ ko.components.register "tf-graph",
...
+      console.log("buckets: ")
+      console.log(buckets)
+      console.log("labels")
+      console.log(['x'].concat(hist_labels))
+      console.log("concat2")
+      
+    
       row_labels = params.row_labels
+      hist_legend = row_labels().slice(1)
+      console.log("row_labels")
```
Comment (tsangjustin) [inline interface/components/graph/index.coffee]: here

File: interface/components/graph/index.coffee [hunk 9]
```diff
@@ -9,28 +9,209 @@ ko.components.register "tf-graph",
...
+
+      row_labels.subscribe ( next ) ->  
         try
-          chart.load
-            xs: getxs()
-            rows: [next].concat data()
+          if global.chart == chart_scatter   
+              chart.load
+                xs: getxs()
+                rows: [next].concat data()
```
Comment (tsangjustin) [inline interface/components/graph/index.coffee]: here

File: interface/components/graph/index.coffee [hunk 10]
```diff
@@ -9,28 +9,209 @@ ko.components.register "tf-graph",
...
+                console.log(data())
+          else if global.chart == chart_histogram
+            console.log("I AM IN SIDE HISTOGRAM NEXT")  
+            console.log("data within next")
+            console.log(data())
+            data_sorted = sort_data_in_next(data())
+            console.log("data_sorted len")
+            console.log(data_sorted.length)
+            if data_sorted.length == 1
+                console.log("buckets in MOTION")
```
Comment (tsangjustin) [inline interface/components/graph/index.coffee]: here

File: interface/components/graph/index.coffee [hunk 11]
```diff
@@ -61,145 +242,62 @@ ko.components.register "tf-graph",
...
-              cross_visible = false
-            else
-              cross_visible = true
-            return
-          if cross_legend
-            cross_legend.onclick = cross_legend_VS
+      global.changeGraph = () ->
+        e = document.getElementById("graphs")
+        selectedvalue = e.options[e.selectedIndex].value
+        console.log("hello, check here")
```
Comment (tsangjustin) [inline interface/components/graph/index.coffee]: remove all these console logs

File: interface/components/graph/index.pug [hunk 12]
```diff
@@ -0,0 +1,4 @@
+tf-graph
```
Comment (tsangjustin) [inline interface/components/graph/index.pug]: Why is this file needed?

**Comment by a311916000** [comment]
console removed, and download button added, and button position fixed.
